lib/bearerstrategy.js

```diff
@@ -118,121 +125,298 @@ const jws = require('jws');
  * @constructor
  */
 
+/*
+ * BearerStrategy used to inherit from passport-http-bearer strategy
+ *     https://github.com/jaredhanson/passport-http-bearer/
+ * when BearerStrategy is created, it reads the info (from developer and metadata).
+ * In the authentication time, passport-http-bearer does all the work using these
+ * info read at the BearerStrategy's creation time.
+ *
+ * This is not good since:
+ * (1) for B2C, different policy may have different metadata, so the metadata should
+ * be loaded dynamically (and cached) in the authentication time based on the request
+ * (2) async issue, metadata loading is async, passport-http-bearer's `authenticate`
+ * function depends on the metadata, and it might be called before the metadata loading
+ * finishes
+ *
+ * Therefore, what we do here is let the metadata loading happen in `authenticate`
+ * function, and use waterfall to make sure the authentication code runs after the
+ * metadata loading is finished. To do this, we write our own `authenticate` function
+ * and no longer inherit passport-http-bearer strategy.
+ */
```

Comment:
Do we need to include what we 'used' to do in the comment?
Perhaps we could just say what we are doing.